Semantics docs

tuzz opened the pull request:
This is now complete and has been reviewed by @tadast and @AndrewVos.

I have addressed their comments.

https://trello.com/c/Zz8AXqBr/369-docs-semantics-of-publishing-api

boffbowsh commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+The `publishing_app` identifies the application that is responsible for
+publishing the content item. When a content item is created, its `base_path` is
+registered to the `publising_app`. The path may not be used by a content item
```
`publising_app`

boffbowsh replied:
(Sorry)

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+The `content_id` is the content item’s main identifier and it forms its identity
+as it travels through the pipeline. This is why requests to create and query
+content items tend to be keyed by `content_id` in the URL of the request.
```
Tend to be?

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+Each `content_id` refers to a single piece of content, with a couple of caveats:
+
+- `content_ids` are shared across locales – the English and French versions of a
```
Technically "`content_id`s"

elliotcm replied:
Or just "Content IDs"

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+a content item share a `content_id`
+
+`content_ids` are [UUIDs](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+and will not be accepted by the Publishing API otherwise.
```
Do we need to specify a format for the content ID?  Do we validate it anywhere other than local tests?

bradwright replied:
govuk-content-schemas [validates that they're GUIDs](https://github.com/alphagov/govuk-content-schemas/blob/44dfad0cc241b7bd9576f0a7cf7f3fdeac8ddfce/formats/metadata.json#L19-L21)

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+and will not be accepted by the Publishing API otherwise.
+
+Note: Previously, the `base_path` was a content item's main identifier. This is
+no longer the case. It has been changed to `content_id` because `base_paths` had
```
Like with content IDs, we can just say "base paths" when referring to the concept rather than the actual field.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+### base_path
+
+Example: */vat-rates*
```
:D

tuzz replied:
:D

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+The `format` specifies the data format of the content item as per the
+[GOV.UK content schemas](https://github.com/alphagov/govuk-content-schemas).
+It is used downstream to render the content item in the manner
+it was intended.
```
"in the manner it was intended" is a bit vague.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+Required: Yes
+
+The `publishing_app` identifies the application that is responsible for
+publishing the content item. When a content item is created, its `base_path` is
```
"the application that published the content item"?

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+The `publishing_app` identifies the application that is responsible for
+publishing the content item. When a content item is created, its `base_path` is
+registered to the `publishing_app`. The path may not be used by a content item
+that was created with a different `publishing_app`.
```
This has now been restricted further.  The path can't be used by any other content ID unless the content item is deleted first or if the new content ID is listed as a `can_be_replaced_by` link.

elliotcm replied:
UPDATE

This is being backed out for gone and redirect items, making `can_be_replaced_by` irrelevant

tuzz replied:
Ok – thanks for the update. I'll leave this as-is for now then.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+registered to the `publishing_app`. The path may not be used by a content item
+that was created with a different `publishing_app`.
+
+The `publishing_app` can then be used to filter content items appropriately when
```
"appropriately"?

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+requests are made to the Publishing API. The `publishing_app` is also used as a
+means of auditing which applications are making use of the publishing pipeline.
+
+Note: `publishing_app` should be hyphenated as this is the convention used in
```
Should make it clear that the value of the field should be hyphenated, not the field itself.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+### public_updated_at
+
+Example: *2015-01-01T12:00:00Z*
```
Do we need to specify a format?  e.g. ISO 8601?

bradwright replied:
We should enforce ISO8601 (if we're not already).

tuzz replied:
Addressed.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+can contain arbitrary JSON that will be stored against the content item.
+
+Not all `formats` have required fields and so details is not required unless the
+`format` demands it. If it is not set, it will default to an empty JSON object.
```
.. as specified in the content schema

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+`routes` appear as a JSON array and each element in this array is a JSON object
+that contains the properties *path* and *type*. No other properties are
+supported.  The *type* must be set to *exact* to denote that the route maps to
+an exact path on the GOV.UK website.
```
Do we still support "prefix"?  If not we should mention that the type is legacy and/or remove it.

tuzz replied:
The code that was ported over from Content store supports 'prefix' types. I don't know whether this is supported in the router, though. @danielroseman ?

bradwright replied:
```
irb(main):005:0> Route.where(route_type: 'prefix').count
=> 3462
```

elliotcm replied:
@bradwright I guess my original question was are we supporting the creation of new ones generally rather than technically.  If so we should update the docs to include them as an option, if not we should mention that they used to be a thing but are no longer.

bradwright replied:
Smart Answers still need prefix routes, as does Trade Tariff.

Confusing Smart Answers [says it needs an exact route](https://github.com/alphagov/smart-answers/blob/9f29165a85dadbd3f905acfb7e6e7fbe5ee9a55e/app/presenters/flow_content_item.rb#L20), but in the router they're all prefixes.

elliotcm replied:
@tuzz Let's update these docs to allow for prefix routes then, but perhaps say they're discouraged in favour of exact routes.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+supported.  The *type* must be set to *exact* to denote that the route maps to
+an exact path on the GOV.UK website.
+
+The `routes` are generally required, except for the case when the content item
```
Remove "generally"

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+has a `format` of *redirect*. In this case, the `routes` must not be present as
+it doesn’t make sense to have routes for a redirect. When the `format` is
+anything except *redirect*, the routes must include the `base_path` of the
+content item, at the very least.
```
"must include" makes "at the very least" redundant.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+An *exact type* denotes that the *path* should be checked for an exact match
+against the user’s request URL when determining whether a redirect should occur.
+A *prefix type* denotes that any subpath under the specified path should be
+redirected to the destination. You can think of these as “exact” and “fuzzy”
```
Maybe wildcard is a more widely understood term than fuzzy?

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+redirected to the destination. You can think of these as “exact” and “fuzzy”
+matches, respectively.
+
+The `redirects` are generally optional, except for the case when the content
```
Remove "generally"

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+The `redirects` are generally optional, except for the case when the content
+item has a `format` of *redirect*. In this case, the redirects must be present
+and they must include the `base_path` of content_item in the *path* property, at
+the very least.
```
- "at the very least"

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+Required: Conditionally
+
+The `update_type` is an indicator of how much the content item has changed since
+the last time it was published. This field is required when publishing the
```
It's more about the importance of the change than the amount.  It could be two words but if it has a significant impact on the readers it would be a major change.  The litmus test is whether the readers would want to be notified about the change or not.

danielroseman commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+There is another document that explains the syntax for how the API should be
+used. That document can be found
+[here](https://docs.google.com/document/d/1zYFZFZ2TLzasgX4SZTxuQhLzqRYCk4hWKBTcb8QFIRA/edit).
```
Is this staying on Google Docs, rather than going into the wiki?

tuzz replied:
Not sure, @steventux ?

steventux replied:
The syntax docs are appended to @elliotcm's overview [on the atlassian wiki](https://gov-uk.atlassian.net/wiki/display/TECH/Publishing+Platform). IMO these would be better placed in GH along with examples and semantics, any objections?

tuzz replied:
No objections. Please could you let me know if you move it so I can update this pull request. Thanks

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+the last time it was published. This field is required when publishing the
+content item.
+
+There’s no restriction on what the `update_type` should be, but at present we
```
Should we restrict this?  The original docs for the content store stated it must be one of the known types which might be added to later.  Three options I see:
1. We place no restrictions and make it a publisher/frontend/bus service concern
2. We place no programmatic restrictions but ask people to talk with us before adding a new type
3. We place programmatic restrictions and deal with type requests as a team

tuzz replied:
I think option 2 sounds the most sensible. I don't think we should be too prescriptive about how update_type should be used as we don't want to be exposed to too much knowledge about downstream consumers.

That being said, I think it's important that our team knows what that field is being used for.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+- *minor* - for minor edits that probably won’t be of interest to users
+
+- *major* - for major edits that probably will be of interest to users
```
Worth mentioning that these trigger email alerts.

rboulton replied:
That may well not remain true in future - in particular, we probably need to send email alerts when links are changed shortly after publication, even when the link change isn't sent as a major update.  I'd suggest adding the word "currently" to this, so that it doesn't look like this is setting a contract that can be relied on.

elliotcm commented:
The fields are grouped well but ordered seemingly randomly.  Alphabetize?

Should we be indicating which fields support govspeak?

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+Required: No
+
+The `needs_ids` are the identifiers of [user needs](https://www.gov.uk/design-principles)
+that are entered through the [Maslow Application](https://github.com/alphagov/maslow).
```
"Maslow application"

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+The `needs_ids` are the identifiers of [user needs](https://www.gov.uk/design-principles)
+that are entered through the [Maslow Application](https://github.com/alphagov/maslow).
+They are passed through to the content store, untouched by the pipeline. The
+front-end applications can then use the `needs_ids` to present pages to users,
```
This sentence doesn't provide any actual information about what type of pages are presented without clicking the link.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+applications will optionally provide this string when querying the content store
+in order to retrieve content items in a given `locale`.
+
+A list of valid locales can be viewed in the Publishing API [here](https://github.com/alphagov/publishing-api/blob/master/config/application.rb#L32-L37).
```
Recommend linking to a blob on a commit rather than master as this may change.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+The `access_limited` field determines who should be granted access to the
+content item before it has been published. At the point of publish the content
+item is public and everyone has access to it. This field should be used when a
+content item needs to be drafted, but it should not be visible on
```
Drop the comma, "it", and mention that it shouldn't be visible on content preview except for authorised users (typically members of the organisation that created the content).

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+### access_limited
+
+Example: *{ users: ["bf3e4b4f-f02d-4658-95a7-df7c74cd0f50"] }*
```
Not 100% sure that user UUIDs follow this format.

tuzz replied:
I took this from the PACT: https://pact-broker.dev.publishing.service.gov.uk/pacts/provider/Content%20Store/consumer/Publishing%20API/latest

elliotcm replied:
I was talking about the ones used by the user model in signon specifically, but it turns out I was mistaken.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+access restriction will be placed on the content item.
+
+When front-end applications make requests to the content store, they must supply
+the user they are making the request on behalf of if the content item is
```
How?  Details of the HTTP header?

tuzz replied:
I'm not sure. @danielroseman, do you know what this HTTP header looks like?

bradwright replied:
https://github.com/alphagov/content-store/pull/129

bradwright replied:
Given that is implemented in the content store rather than in the Publishing API, I'd be tempted to either link to the content store OR say that "items with access_limited are limited to only the user IDs appearing in the list" (since the details of the content-store implementation may change).

tuzz replied:
:+1: I decided to link to that pull request. I don't think it's likely to change any time soon and we'll just have to be mindful to update these docs if they do.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+published.
+
+An update to the `links` causes a message to be placed on the message queue.
+This message will have a special update_type of links (see [**update_type**](#update_type)).
```
NB no emails sent

tuzz replied:
Addressed.

elliotcm commented on doc/semantics-of-the-publishing-api.md:
```diff
@@ -0,0 +1,406 @@
+
+An update to the `links` causes a message to be placed on the message queue.
+This message will have a special update_type of links (see [**update_type**](#update_type)).
+This queue is consumed by the Rummager application in order to reindex the
```
Is it?

tuzz replied:
Not sure. @danielroseman, @steventux ?

tuzz commented:
> @elliotcm: The fields are grouped well but ordered seemingly randomly. Alphabetize?

Done. I decided to keep routes/redirects together rather than wedging rendering_app in between. I also thought that it makes more sense to put routes before redirects.

tuzz commented:
> @elliotcm: Should we be indicating which fields support govspeak?

I've added this as a "Post merge" step on [that pull request](https://github.com/alphagov/publishing-api/pull/142). I'll document it then.

tuzz commented:
I have addressed your comments as much as possible @elliotcm. There are a few questions that I have CC'ed other people into (@danielroseman, @bradwright, @steventux).

Thanks

tuzz commented:
I have addressed all comments again. @elliotcm @danielroseman @steventux @bradwright 

Thanks

bradwright commented:
For my part this looks good :+1:
